File: src/test/java/org/trp/shincolle/entity/base/ShipFollowDecisionResolverTest.java
package org.trp.shincolle.entity.base;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

class ShipFollowDecisionResolverTest {

    @Test
    void followTeleportShouldIgnoreNearbyNonForcedTargets() {
        ShipMovementRecoveryState recovery = new ShipMovementRecoveryState();

        assertFalse(ShipFollowDecisionResolver.shouldTryTeleport(
                new ShipFollowDecisionResolver.State(recovery, false, 64.0D)));
    }

    @Test
    void followTeleportShouldRespectForcedCooldown() {
        ShipMovementRecoveryState recovery = new ShipMovementRecoveryState();
        ShipFollowDecisionResolver.State state =
                new ShipFollowDecisionResolver.State(recovery, true, 1.0D);

        assertTrue(ShipFollowDecisionResolver.shouldTryTeleport(state));
        assertFalse(ShipFollowDecisionResolver.shouldTryTeleport(state));
        assertFalse(ShipFollowDecisionResolver.shouldTryTeleport(state));
        for (int i = 0; i < ShipAiNumbers.FOLLOW_TELEPORT_COOLDOWN_TICKS - 2; i++) {
            assertFalse(ShipFollowDecisionResolver.shouldTryTeleport(state));
        }
        assertTrue(ShipFollowDecisionResolver.shouldTryTeleport(state));
    }
}

File: src/test/kotlin/org/trp/shincolle/entity/base/ShipFollowDecisionResolverTest.kt
package org.trp.shincolle.entity.base

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

class ShipFollowDecisionResolverTest {

    @Test
    fun `follow teleport should ignore nearby non forced targets`() {
        val recovery = ShipMovementRecoveryState()

        assertThat(
            ShipFollowDecisionResolver.shouldTryTeleport(
                ShipFollowDecisionResolver.State(recovery, false, 64.0)
            )
        ).isFalse()
    }

    @Test
    fun `follow teleport should respect forced cooldown`() {
        val recovery = ShipMovementRecoveryState()
        val state = ShipFollowDecisionResolver.State(recovery, true, 1.0)

        assertThat(ShipFollowDecisionResolver.shouldTryTeleport(state)).isTrue()
        assertThat(ShipFollowDecisionResolver.shouldTryTeleport(state)).isFalse()
        assertThat(ShipFollowDecisionResolver.shouldTryTeleport(state)).isFalse()
        repeat(ShipAiNumbers.FOLLOW_TELEPORT_COOLDOWN_TICKS - 2) {
            assertThat(ShipFollowDecisionResolver.shouldTryTeleport(state)).isFalse()
        }
        assertThat(ShipFollowDecisionResolver.shouldTryTeleport(state)).isTrue()
    }
}
